parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +628,349 @@ protected GroupType build(String name) {
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public MapBuilder(P parent) {
+      super(parent);
+    }
+
+    private MapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected MapBuilder<P> self() {
+      return this;
+    }
+
+    public MapKeyBuilder<P> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P>(this, type);
+      } else {
+        return new MapKeyBuilder<P>(this, returnType, type);
+      }
+    }
+
+    public MapGroupKeyBuilder<P> groupKey() {
+      return new MapGroupKeyBuilder<P>(this);
+    }
+
+    @Override
+    protected Type build(String name) {
+      if (keyType == null) {
+        keyType = STRING_KEY;
+      }
+      return buildGroup(repetition)
+          .as(OriginalType.MAP)
+          .repeatedGroup()
+              .addFields(keyType, valueType)
+              .named("map")
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a MAP type.");
+      return super.named(name);
+    }
+  }
+
+  public static class ListBuilder<P> extends Builder<ListBuilder<P>, P> {
+    private Type elementType = null;
+
+    public ListBuilder(P parent) {
+      super(parent);
+    }
+
+    public ListBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected ListBuilder<P> self() {
+      return this;
+    }
+
+    @Override
+    protected Type build(String name) {
+      Preconditions.checkNotNull(elementType, "List element type");
+      return buildGroup(repetition)
+          .as(OriginalType.LIST)
+          .addField(repeatedGroup()
+              .addFields(elementType)
+              .named("list"))
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a LIST type.");
+      return super.named(name);
+    }
+
+    public GroupBuilder<ListBuilder<P>> groupElement(Type.Repetition repetition) {
+      Preconditions.checkState(elementType == null, "groupElement can be called just once on a " +
+          "ListBuilder");
+      return new GroupBuilder<ListBuilder<P>>(this).repetition
+          (repetition);
+    }
+
+    public GroupBuilder<ListBuilder<P>> requiredGroupElement() {
+      return groupElement(Type.Repetition.REQUIRED);
+    }
+
+    public GroupBuilder<ListBuilder<P>> optionalGroupElement() {
+      return groupElement(Type.Repetition.OPTIONAL);
+    }
+
+    public PrimitiveBuilder<ListBuilder<P>> required(PrimitiveTypeName type) {
+      Preconditions.checkState(elementType == null, "groupElement can be called just once on a " +
+          "ListBuilder");
+      return new PrimitiveBuilder<ListBuilder<P>>(this, type).repetition(Type.Repetition
```

Comment:
Do you mean name should be PrimitiveElementBuilder, instead of required? I wanted to keep it similar to existing public methods. Maybe I am missing something here.